pandas/tests/frame/test_alter_axes.py

```diff
@@ -270,6 +270,31 @@ def test_set_index_raise_on_type(self, frame_of_index_cols, box,
             df.set_index(['A', df['A'], box(df['A'])],
                          drop=drop, append=append)
 
+    def test_set_index_custom_label_type(self):
+        # GH 24969
+
+        class Thing(object):
+            def __init__(self, name, color):
+                self.name = name
+                self.color = color
+
+            def __str__(self):
+                return "<Thing %r>" % (self.name,)
+
+        thing1 = Thing('One', 'red')
+        thing2 = Thing('Two', 'blue')
+        df = DataFrame({thing1: [0, 1], thing2: [2, 3]})
+        expected = DataFrame({thing1: [0, 1]},
+                             index=Index([2, 3], name=thing2))
```

Comment:
I guess can you add an object that is not hashable (maybe just inherit an object from list) and see what happens

Reply:
To test what? Failure?
Lists would currently be interpreted as array-likes.

I am already testing sets, which are not hashable.

Reply:
no thus is a test of a custom class that fails hashing for example

Reply:
But you cannot create a dataframe with an unhashable object as column name in the first place, so you would actually be testing that you are passing an object that is not in the column names, which is already covered I think?

Reply:
That's what I thought, but...

```python
df = pd.DataFrame([[0, 1]], columns=[{'a'}, {'b'}])
```

though essentially everything is broken on this dataframe :).

Reply:
So then I would say we shouldn't be testing that, but rather open an issue about it ? :-)

Although I suppose we simply don't test hashability of object-dtype ndarrays at Index creation time, but only when it is used. Which makes sense, as I suppose it can be quite costly to do so.

Reply:
Adding a custom class that is unhashable is such a pathological edge case that it really does not make sense to add a test. As @TomAugspurger showed, the constructor essentially goes through, but then almost everything else is broken, even the repr.

Unhashable types (in this case `set` and `iter`) are already tested to raise an error.

Reply:
Ok, I found a way to still reliably test this. But then the question if `set` should raise a `ValueError` as currently, or a `TypeError` as something that's not hashable. I went with the latter, as it would be the only consistent treatment for what has been specified by @jreback above (i.e. `TypeError` for unhashable)